File: polyglot/Filter/filter.cs
// Filter design pattern
using System;
using System.Collections.Generic;
using System.Linq;  // Where method

using PersonList = System.Collections.Generic.IEnumerable<Person>;

// This pattern allows to use a list of objects and perform
// a filtering operation on that list so as to obtain a new
// list comprised of those objects in the initial list which
// satisfy a given predicate. Since the introduction of
// lambda expressions within Java 8 and the introduction
// of functional interfaces such as Predicate<T>, this 
// pattern is not very useful in practice and amounts 
// mainly to a coding exercise aiming at re-implementing
// the Predicate<T> java interface. This pattern is not
// useful either in functional languages which directly 
// support first class functions and filter operations on
// lists, or in C# which supports Where on IEnumerable

public delegate bool PredicateFunc<T>(T t);

class Predicate<T> {
  private readonly PredicateFunc<T> test;
  public Predicate(PredicateFunc<T> test){ this.test = test; }
  public bool Test(T t){ return test(t); }

  public static Predicate<T> IsEqual(T targetRef){
    return new Predicate<T>(t => t.Equals(targetRef));
  }

  public Predicate<T> Not(){
    return new Predicate<T>(t => !Test(t));
  }

  public Predicate<T> And(Predicate<T> other){
    return new Predicate<T>(t => !Test(t) ? false : other.Test(t));
  }

  public Predicate<T> Or(Predicate<T> other){
    return new Predicate<T>(t => Test(t) ? true : other.Test(t));
  }
}

class Person {

  private readonly string name;
  private readonly string gender;
  private readonly string maritalStatus;

  public Person(string name, string gender, string maritalStatus){
    this.name = name;
    this.gender = gender;
    this.maritalStatus = maritalStatus;
  }

  public string Name            {get{return name;}}
  public string Gender          {get{return gender;}}
  public string MaritalStatus   {get{return maritalStatus;}}

  // check list of 'Equals' overriding:
  // 1. having the right signature: bool Equals(object other)
  // 2. accompanying override for GetHashCode: 
  // x == y => x.GetHashCode == y.GetHashCode
  // 3. It is not defined in terms of mutable fields
  // 4. It defines an equivalence relation
  public override bool Equals(object other){  // possibly equality
    if(other is Person){
      return Name.Equals(((Person) other).Name);
    } else {
      return false;
    }
  }
  // it is a sin to override 'Equals' without overriding 'GetHashCode'
  public override int GetHashCode(){
    return Name.ToUpper().GetHashCode();
  }


  public static readonly Predicate<Person> male =
    new Predicate<Person>(
        t => t.Gender.Equals("MALE",StringComparison.OrdinalIgnoreCase)
    );

  public static readonly Predicate<Person> female =
    new Predicate<Person>(
        t => t.Gender.Equals("FEMALE",StringComparison.OrdinalIgnoreCase)
    );

  public static readonly Predicate<Person> single =
    new Predicate<Person>(
        t => t.MaritalStatus.Equals("SINGLE",StringComparison.OrdinalIgnoreCase)
    );

  public static readonly Predicate<Person> singleMale = single.And(male);
  public static readonly Predicate<Person> singleOrFemale = single.Or(female);


  public static PersonList People(){
    IList<Person> people = new List<Person>();
    people.Add(new Person("Robert","Male","Single"));
    people.Add(new Person("John","Male","Married"));
    people.Add(new Person("Laura","Female","Married"));
    people.Add(new Person("Diana","Female","Single"));
    people.Add(new Person("Mike","Male","Single"));
    people.Add(new Person("Bobby","Male","Single"));

    return people as PersonList;  // explicit upcast
  }

  public static void Print(PersonList list){
    foreach(Person person in list){
      Console.Write("(" + person.Name
          + "," + person.Gender
          + "," + person.MaritalStatus
          + ")\t");
    }
  }

  public static PersonList filter(PersonList list, Predicate<Person> pred){
    return list.Where(p => pred.Test(p));
  }
}

public class Filter {
  public static void Main(string[] args){
    Person john2 = new Person("John","Male","Married");
    Predicate<Person> notJohn = Predicate<Person>.IsEqual(john2).Not();

    PersonList people          = Person.People();
    PersonList males           = Person.filter(people, Person.male);
    PersonList females         = Person.filter(people, Person.female);
    PersonList singleMales     = Person.filter(people, Person.singleMale);
    PersonList singleOrFemales = Person.filter(people, Person.singleOrFemale);
    PersonList notJohns        = Person.filter(people, notJohn);

    Console.Write("Everyone:\t\t");         Person.Print(people);    
    Console.Write("\nNot John:\t\t");       Person.Print(notJohns);    
    Console.Write("\nSingle or Female:\t"); Person.Print(singleOrFemales);    
    Console.Write("\nMales:\t\t\t");        Person.Print(males);    
    Console.Write("\nSingle Males:\t\t");   Person.Print(singleMales);    
    Console.Write("\nFemales:\t\t");        Person.Print(females);    
    Console.Write("\n");
  }
}

File: polyglot/Filter/Filter.scala
// Filter design pattern

// This pattern allows to use a list of objects and perform
// a filtering operation on that list so as to obtain a new
// list comprised of those objects in the initial list which
// satisfy a given predicate. Since the introduction of
// lambda expressions within Java 8 and the introduction
// of functional interfaces such as Predicate<T>, this 
// pattern is not very useful in practice and amounts 
// mainly to a coding exercise aiming at re-implementing
// the Predicate<T> java interface. This pattern is not
// useful either in functional languages which directly 
// support first class functions and filter operations on lists.

class Predicate[A](closure: A => Boolean) {
  def apply(a: A) = closure(a)
  def unary_!() = new Predicate[A](a => !this(a))
  def &(other: Predicate[A]) = 
    new Predicate[A](a => if(!this(a)) false else other(a))
  def |(other: Predicate[A]) = 
    new Predicate[A](a => if(this(a)) true else other(a))
}

object Predicate {
  def isEqual[A](targetRef: A): Predicate[A] = new Predicate(_ == targetRef)
}

class Person(val _name: String, val _gender: String, val _maritalStatus: String){

  // There is no real need to define getters, but here we go
  def name          = _name
  def gender        = _gender
  def maritalStatus = _maritalStatus

  // equals used by operator == 
  // check list of 'equals' overriding:
  // 1. having the right signature equals(other: Any): Boolean
  // 2. accompanying override for hashCode: x == y => x.hashCode == y.hashCode
  // 3. It is not defined in terms of mutable fields
  // 4. It defines an equivalence relation
  override def equals(other: Any) = other match {
    case p: Person  => name.equalsIgnoreCase(p.name)
    case _          => false
  } 
  // it is a sin to override 'equals' without overriding 'hashCode'
  override def hashCode = name.toUpperCase.hashCode  
  
  override def toString() = "("+ name +","+ gender +","+ maritalStatus +")"
}

// statics
object Person {

  // some static predicates
  val male    = new Predicate[Person](_.gender.equalsIgnoreCase("MALE"))
  val female  = new Predicate[Person](_.gender.equalsIgnoreCase("FEMALE"))
  val single  = new Predicate[Person](_.maritalStatus.equalsIgnoreCase("SINGLE"))
  val singleMale      = single & male
  val singleOrFemale  = single | female

  // sample of known persons
  def people: List[Person] =  {
    new Person("Robert","Male","Single")  ::
    new Person("John","Male","Married")   ::
    new Person("Laura","Female","Married")::
    new Person("Diana","Female","Single") ::
    new Person("Mike","Male","Single")    ::
    new Person("Bobby","Male","Single")   :: Nil
  }

  // printing list of people
  def printList(list: List[Person]) = {
    list.map(person => print(person + "\t"))
  }

  def filterList(list:List[Person], predicate:Predicate[Person]):List[Person] = {
    if(predicate == null) return list
    list.filter(predicate(_))
  }
}

object Filter {

  def main(args: Array[String]){
    val john2 = new Person("John","Male","Married")
    val notJohn: Predicate[Person] =  !Predicate.isEqual(john2)
    
    val people          = Person.people
    val males           = Person.filterList(people, Person.male)
    val females         = Person.filterList(people, Person.female)
    val singleMales     = Person.filterList(people, Person.singleMale)
    val singleOrFemales = Person.filterList(people, Person.singleOrFemale)
    val notJohns        = Person.filterList(people, notJohn)
    
    print("Everyone:\t\t");         Person.printList(people)
    print("\nNot John:\t\t");       Person.printList(notJohns)
    print("\nSingle or Female:\t"); Person.printList(singleOrFemales)
    print("\nMales:\t\t\t");        Person.printList(males)
    print("\nSingle Males:\t\t");   Person.printList(singleMales)
    print("\nFemales:\t\t");        Person.printList(females)
    print("\n")
  }
}

